src/category_theory/limits/shapes/zero.lean

```diff
@@ -0,0 +1,93 @@
+/-
+Copyright (c) 2019 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import category_theory.limits.shapes.terminal
+
+/-!
+# Zero morphisms and zero objects
+
+A category "has zero morphisms" if there is a designated "zero morphism" in each morphism space,
+and compositions of zero morphisms with anything give the zero morphism. (Notice this is extra
+structure, not merely a property.)
+
+A category "has a zero object" if it has an object which is both initial and terminal. Having a
+zero object provides zero morphisms, as the unique morphisms factoring through the zero object.
+
+## References
+* https://en.wikipedia.org/wiki/Zero_morphism
+-/
+
+universes v u
+
+open category_theory
+
+namespace category_theory.limits
+
+variables (C : Type u) [𝒞 : category.{v} C]
+include 𝒞
+
+/-- A category "has zero morphisms" if there is a designated "zero morphism" in each morphism space,
+and compositions of zero morphisms with anything give the zero morphism. -/
+class has_zero_morphisms :=
+[has_zero : Π X Y : C, has_zero (X ⟶ Y)]
+(comp_zero' : ∀ {X Y : C} (f : X ⟶ Y) (Z : C), f ≫ (0 : Y ⟶ Z) = (0 : X ⟶ Z) . obviously)
+(zero_comp' : ∀ (X : C) {Y Z : C} (f : Y ⟶ Z), (0 : X ⟶ Y) ≫ f = (0 : X ⟶ Z) . obviously)
+
+attribute [instance] has_zero_morphisms.has_zero
+restate_axiom has_zero_morphisms.comp_zero'
+attribute [simp] has_zero_morphisms.comp_zero
+restate_axiom has_zero_morphisms.zero_comp'
+attribute [simp, reassoc] has_zero_morphisms.zero_comp
+
+/-- A category "has a zero object" if it has an object which is both initial and terminal. -/
+class has_zero_object :=
+(zero : C)
```

Comment:
Any reason to treat this `zero` differently than `has_zero` in `has_zero_morphisms`? (But see below.)